clusterloader2/testing/measurements-gatherer/config.yaml

```diff
@@ -0,0 +1,176 @@
+#Constants
+{{$LARGE_BACKEND_SIZE := 300}}
+{{$MEDIUM_BACKEND_SIZE := 150}}
+{{$SMALL_BACKEND_SIZE := 10}}
+{{$LARGE_BACKEND_LABEL := "ilb-large"}}
+{{$MEDIUM_BACKEND_LABEL := "ilb-medium"}}
+{{$SMALL_BACKEND_LABEL := "ilb-small"}}
+# adding a fixed value for first version of the test, rate of pod creation not a concern yet.
+{{$ilbQPS := 500}}
+{{$namespaces := 1}}
+
+#Test
+name: ilbload
+automanagedNamespaces: {{$namespaces}}
+tuningSets:
+- name: ILBConstantQPS
+  qpsLoad:
+    averageQps: {{$ilbQPS}}
+steps:
+# Mesure each of the ILB services separately, this will provide insight on how long programming
+# ILB takes as a function of number of backends.
+- measurements:
+  #- Identifier: ServiceCreationLatencyLarge
+  #  Method: ServiceCreationLatency
+  #  Params:
+  #    action: start
+  #    labelSelector: size = {{$LARGE_BACKEND_LABEL}}
+  #- Identifier: ServiceCreationLatencyMedium
+  #  Method: ServiceCreationLatency
+  #  Params:
+  #    action: start
+  #    labelSelector: size = {{$MEDIUM_BACKEND_LABEL}}
+  - Identifier: CreateServices
+    Method: GatherMeasurements
+    Params:
+      Identifiers: ServiceCreationLatencyLarge, ServiceCreationLatencySmall, ServiceCreationLatencyMedium
+      Method: ServiceCreationLatency
+      action: start
+      labelSelector: size = {{$SMALL_BACKEND_LABEL}}, size = {{$MEDIUM_BACKEND_LABEL}}, size = {{$LARGE_BACKEND_LABEL}}
```

Comment:
Instead of having this separated by commas, can this be a repeated field (similarly for Identifiers probably), i.e.:
labelSelector:
- size = {$SMALL...}
- size = ...

+ some validation that the number of items in all fields is equal

Reply:
The problem could be that some of the configs can be common. like action. the way I have implemented currently, it's easier to figure out if its common one or not. 
I had thought about validation but the confusion I had was what to do if the number of items in all fields are equal or not. Should i fail the entire measurement execution ?

Reply:
I'd extend Wojtek's idea a bit. Ideally, I'd like to have something like this (+- naming)

```
- Identifier: CreateServices
  Method: GatherMeasurements
  Params:
    Identifiers: 
      ServiceCreationLatencyLarge:
         labelSelector: size = {{$SMALL_BACKEND_LABEL}}
      ServiceCreationLatencySmall:
         labelSelector: size = {{$MEDIUM_BACKEND_LABEL}}
      ServiceCreationLatencyMedium:
         labelSelector: size = {{$LARGE_BACKEND_LABEL}}
         someExtraParam: onlyForMedium
    Method: ServiceCreationLatency
    action: start
    common_param: xxx
```

So basically the Indentifiers will be a map, from measurement name to custom parameters for this particular measurement. The common params will be stored in the top level Params stanza.

If there are no custom parameters, providing an empty map should work, e.g.:
```
- Identifier: GatherServices
  Method: GatherMeasurements
  Params:
    Identifiers: 
      ServiceCreationLatencyLarge:
      ServiceCreationLatencySmall:
      ServiceCreationLatencyMedium:
    Method: ServiceCreationLatency
    action: gather
```

WDYT?

Also, could you please rebase the PR now the MeasurementManager change has merged?

Reply:
Yeah this makes sense. I will integrate these changes and also rebase the PR. Thanks!

Reply:
Yeah - even better.

Reply:
@mm4tt I have made the changes. While testing I noticed a couple of issues:
1) For some reason. the service creation latency measurement throws a segmentation violation error at this line:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/service_creation_latency.go#L316
I debugged further and found that the pod object is being returned as nil here:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/service_creation_latency.go#L306 

which is because the podservice is nil:

https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/execservice/exec_service.go#L137

This could be an issue with my setup, but what I noticed was that in service_creation_latency, we aren't handling the err returned by execservice.GetPod(). Could that be an issue?

Also, this is something which has suddenly stopped working and was fine till last week.

2) The other thing is that the errorlist size is always initialized to 2 here:

https://github.com/kubernetes/perf-tests/blob/b2afa608d8fd3134de6fad15350d469054c6f5f8/clusterloader2/pkg/measurement/common/gather_measurements.go#L60

even though I am creating it there. Because of that, this warning keeps getting printed:
https://github.com/kubernetes/perf-tests/blob/b2afa608d8fd3134de6fad15350d469054c6f5f8/clusterloader2/pkg/test/simple_test_executor.go#L152 but the errList is empty. Have you noticed something like this before?

Reply:
Re. 1
This is a clear bug in our code. Thanks for debugging. It sounds like an easy one to fix (error checking + don't run service_creation_latency measurement if execservice is disabled. Would you mind fixing it in a separate PR or opening an issue for that?

Re. 2

I haven't seen anything like that. NewErrorList should return an empty list. Are you sure that no error is added to the list anywhere later? What exact warning gets printed, i.e. what is the content of the errList being reported?

Reply:
hey @mm4tt , regarding 1, I have created this issue to track this problem:
https://github.com/kubernetes/perf-tests/issues/1101 I have assigned it to myself. I will take it up. 

2) I had created the error list as the first line of the Execute method and still got this. Also, the errList that gets printed shows up as []

Reply:
1- done.
2- This error now seems to have gone away. I am returning nil now if the error list is empty. Since simple_test_executor checks for nil type and since we are always creating a new error list, it wasn't nil and hence was getting printed.